File: maven-confluence-core/src/test/java/org/bsc/confluence/model/SiteLoadTest.java
package org.bsc.confluence.model;

import static org.hamcrest.core.IsEqual.equalTo;
import static org.hamcrest.core.IsNull.notNullValue;
import static org.junit.Assert.assertThat;

import java.io.InputStream;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.List;
import java.util.function.Consumer;

import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.ExpectedException;

import com.fasterxml.jackson.databind.JsonMappingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.dataformat.yaml.YAMLFactory;

public class SiteLoadTest {

    @Rule
    public ExpectedException thrown = ExpectedException.none();

    private void loadFromYAML( String resource, Consumer<Site> c ) throws Exception {
        final ObjectMapper mapper = new ObjectMapper(new YAMLFactory());
        
        try( InputStream is = getClass().getClassLoader().getResourceAsStream(resource) ) {
            Site site = mapper.readValue( is, Site.class  );
            
            assertThat( site, notNullValue());
            
            c.accept(site);
        }
    
    }
    
    @Test
    public void testIssue182() throws Exception {
    
        //thrown.expect(Exception.class);
        thrown.expect(JsonMappingException.class);
        //thrown.expectMessage("name [page 1 ] is not valid!");
        loadFromYAML( "site-issue182.yaml", site -> {
                        
        });

    }
    
    @Test
    public void testLoadFromYAML() throws Exception {
        
        loadFromYAML( "site.yaml", site -> {
            
            final Path basedir = Paths.get("/tmp");
            site.setBasedir( basedir );
            
            assertThat( site.getHome(), notNullValue());
            assertThat( site.getHome().getUri(), equalTo( 
                    Paths.get(basedir.toString(), "encoding.confluence").toUri()));
            
            final List<Site.Page> children = site.getHome().getChildren();
            
            assertThat( children, notNullValue());
            assertThat( children.size(), equalTo( 2 ) );

            final List<Site.Attachment> attachments = site.getHome().getAttachments();
            
            assertThat( attachments, notNullValue());
            assertThat( attachments.size(), equalTo( 1 ) );
            
        });
            
    }
}

File: core/src/test/kotlin/org/bsc/confluence/model/SiteLoadTest.kt
package org.bsc.confluence.model

import com.fasterxml.jackson.databind.DeserializationFeature
import com.fasterxml.jackson.databind.JsonMappingException
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.databind.json.JsonMapper
import com.fasterxml.jackson.dataformat.xml.XmlMapper
import com.fasterxml.jackson.dataformat.yaml.YAMLFactory
import com.fasterxml.jackson.module.jaxb.JaxbAnnotationModule
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test
import java.nio.file.Files
import java.nio.file.Paths


class SiteLoadTest {

    private fun loadFromYAML(resource: String, c: ( site:Site ) -> Unit = {})  {
        val mapper = ObjectMapper(YAMLFactory())
        javaClass.classLoader.getResourceAsStream(resource).use { `is` ->
            val site = mapper.readValue(`is`, Site::class.java)
            assertNotNull(site)
            c(site)
        }
    }
    private fun loadFromXML(resource: String, c: ( site:Site ) -> Unit = {})  {
        val mapper = XmlMapper() //ObjectMapper(XmlFactory())
        mapper.disable(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES)
        mapper.enable(DeserializationFeature.ACCEPT_SINGLE_VALUE_AS_ARRAY);
        mapper.registerModule(JaxbAnnotationModule())

        javaClass.classLoader.getResourceAsStream(resource).use { `is` ->
            val site = mapper.readValue(`is`, Site::class.java)
            assertNotNull(site)
            c(site)
        }
    }

    private fun loadFromJSON(resource: String, c: ( site:Site ) -> Unit = {})  {
        val mapper = JsonMapper() //ObjectMapper(XmlFactory())
        javaClass.classLoader.getResourceAsStream(resource).use { `is` ->
            val site = mapper.readValue(`is`, Site::class.java)
            assertNotNull(site)
            c(site)
        }
    }

    @Test
    fun testIssue182() = assertThrows(JsonMappingException::class.java) {
            loadFromYAML("site-issue182.yaml")
    }

    @Test
    fun testLoadFromXML() {
        val tempDirectory = Files.createTempDirectory(null)

        loadFromXML("site.xml") { site ->

            val basedir = Paths.get(tempDirectory.toString())
            site.basedir = basedir

            assertNotNull(site.home)
            val uri = Paths.get(basedir.toString(), "encoding.confluence").toUri()
            assertEquals( uri, site.home.uri)
            val children = site.home.children
            assertNotNull(children)
            assertEquals(2, children.size)
            val attachments = site.home.attachments
            assertNotNull(attachments)
            assertEquals(1, attachments.size)
            val labels = site.labels
            assertNotNull(labels)
            assertEquals(2, labels.size)
        }
    }

    @Test
    fun testLoadFromYAML() {
        val tempDirectory = Files.createTempDirectory(null)

        loadFromYAML("site.yaml") { site ->

            val basedir = Paths.get(tempDirectory.toString())
            site.basedir = basedir

            assertNotNull(site.home)
            val uri = Paths.get(basedir.toString(), "encoding.confluence").toUri()
            assertEquals( uri, site.home.uri)
            val children = site.home.children
            assertNotNull(children)
            assertEquals(2, children.size)
            val attachments = site.home.attachments
            assertNotNull(attachments)
            assertEquals(1, attachments.size)
            val labels = site.labels
            assertNotNull(labels)
            assertEquals(2, labels.size)

        }
    }

    @Test
    fun testLoadFromJSON() {
        val tempDirectory = Files.createTempDirectory(null)

        loadFromJSON("site.json") { site ->

            val basedir = Paths.get(tempDirectory.toString())
            site.basedir = basedir

            assertNotNull(site.home)
            val uri = Paths.get(basedir.toString(), "encoding.confluence").toUri()
            assertEquals( uri, site.home.uri)
            val children = site.home.children
            assertNotNull(children)
            assertEquals(2, children.size)
            val attachments = site.home.attachments
            assertNotNull(attachments)
            assertEquals(1, attachments.size)
            val labels = site.labels
            assertNotNull(labels)
            assertEquals(2, labels.size)
        }
    }
}